// ngx-masonry/esm2015/lib/ngx-masonry.module.js
import * as tslib_1 from "tslib";
import { NgModule } from '@angular/core';
import { NgxMasonryComponent } from './ngx-masonry.component';
import { NgxMasonryDirective } from './ngx-masonry.directive';
import * as ɵngcc0 from '@angular/core';
let NgxMasonryModule = class NgxMasonryModule {
}
NgxMasonryModule.ngModuleDef = ɵngcc0.ɵɵdefineNgModule({ type: NgxMasonryModule });
/*@__PURE__*/ ɵngcc0.ɵɵsetNgModuleScope(NgxMasonryModule, { declarations: function () { return [NgxMasonryComponent, NgxMasonryDirective]; }, exports: function () { return [NgxMasonryComponent, NgxMasonryDirective]; } });
/*@__PURE__*/ ɵngcc0.ɵsetClassMetadata(NgxMasonryModule, [{
        type: NgModule,
        args: [{
                imports: [],
                declarations: [NgxMasonryComponent, NgxMasonryDirective],
                exports: [NgxMasonryComponent, NgxMasonryDirective]
            }]
    }], null, null);
NgxMasonryModule.ngInjectorDef = ɵngcc0.ɵɵdefineInjector({ factory: function NgxMasonryModule_Factory(t) { return new (t || NgxMasonryModule)(); }, imports: [[]] });;
export { NgxMasonryModule };
