// ngx-masonry-gallery/esm2015/lib/masonry-gallery.module.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { MasonryGalleryComponent } from './masonry-gallery.component';
import * as ɵngcc0 from '@angular/core';
export class MasonryGalleryModule {
}
MasonryGalleryModule.ngModuleDef = ɵngcc0.ɵɵdefineNgModule({ type: MasonryGalleryModule });
/*@__PURE__*/ ɵngcc0.ɵɵsetNgModuleScope(MasonryGalleryModule, { declarations: function () { return [MasonryGalleryComponent]; }, imports: function () { return [CommonModule]; }, exports: function () { return [MasonryGalleryComponent]; } });
/*@__PURE__*/ ɵngcc0.ɵsetClassMetadata(MasonryGalleryModule, [{
        type: NgModule,
        args: [{
                imports: [
                    CommonModule
                ],
                declarations: [
                    MasonryGalleryComponent,
                ],
                exports: [
                    MasonryGalleryComponent,
                ]
            }]
    }], null, null);
MasonryGalleryModule.ngInjectorDef = ɵngcc0.ɵɵdefineInjector({ factory: function MasonryGalleryModule_Factory(t) { return new (t || MasonryGalleryModule)(); }, imports: [[
            CommonModule
        ]] });
